source/core/coverage/ut_coverage.pkb

```diff
@@ -0,0 +1,135 @@
+  begin
+    g_skipped_objects := ut_object_names();
+    if a_schema_names is not null and a_schema_names.count>0 then
+      g_schema_names := a_schema_names;
+    else
+      g_schema_names := ut_varchar2_list(sys_context('userenv','current_schema'));
+    end if;
+  end;
+
+  function coverage_start(a_schema_names ut_varchar2_list := ut_varchar2_list(sys_context('userenv','current_schema'))) return integer is
+  begin
+    set_schema_names(a_schema_names);
+    return ut_coverage_helper.coverage_start('utPLSQL Code coverage run '||ut_utils.to_string(systimestamp));
+  end;
+
+  procedure coverage_start(a_schema_names ut_varchar2_list := ut_varchar2_list(sys_context('userenv','current_schema'))) is
+    l_coverage_id integer;
+  begin
+    l_coverage_id := coverage_start(a_schema_names);
+  end;
+
+  procedure coverage_start_develop is
+  begin
+    set_schema_names(null);
+    ut_coverage_helper.coverage_start_develop();
+  end;
+
+  procedure coverage_flush is
+  begin
+    ut_coverage_helper.coverage_flush();
+  end;
+
+  procedure coverage_pause is
+  begin
+    ut_coverage_helper.coverage_pause();
+  end;
+
+  procedure coverage_resume is
+  begin
+    ut_coverage_helper.coverage_resume();
+  end;
+
+  procedure coverage_stop is
+  begin
+    ut_coverage_helper.coverage_resume();
+  end;
+
+  procedure skip_coverage_for(a_object ut_object_name) is
+  begin
+    if g_skipped_objects is null then
+      g_skipped_objects := ut_object_names();
+    end if;
+      g_skipped_objects.extend;
+      g_skipped_objects(g_skipped_objects.last) := a_object;
+  end;
+
+  function get_coverage_data(a_coverage_id integer := null) return t_coverage is
+
+    type t_coverage_row is record(
+      name          varchar2(500),
+      line_number   integer,
+      total_occur   number(38,0)
+    );
+    type tt_coverage_rows is table of t_coverage_row;
+
+    l_data             ut_coverage_rows;
+    l_result           t_coverage;
+    l_new_unit         t_unit_coverage;
+    l_skipped_objects  ut_object_names := ut_object_names();
+  begin
+
+    if not ut_coverage_helper.is_develop_mode() then
+      l_skipped_objects := ut_utils.get_utplsql_objects_list() multiset union set(g_skipped_objects);
+    end if;
+
+    -- TODO - add inclusive and exclusive filtering
+    select ut3.ut_coverage_row(
+             lower(s.owner||'.'||s.name),
+             s.owner, s.name, s.type, s.line,
+           --filtering out false - negatives reported by profiler (zero executions while it should be ignored).
+           case when
+              regexp_instr(
+                s.text,
+                '^\s*(((not)?\s*(overriding|final|instantiable)\s*)*(constructor|member)?\s*(procedure|function)|begin|end\s*;)', 1, 1, 0, 'i'
+              ) = 0 then u.total_occur
+           end
+          )
+      bulk collect into l_data
+      from all_source s
+      --need to use owner privileges here
+      left join table( ut3.ut_coverage_helper.get_raw_coverage_data() ) u
```

Comment:
It's a question of:
- do we do it in one big shot or do we want to do it row by row (object by object)?
- object by object might generate many individual queries to dba_source - but this actually might be faster than one big join ans as you said - there is a risg of large data volumes getting generated.